docs/proposals/deployment.md

```diff
@@ -0,0 +1,232 @@
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest 1.0.x release of this document can be found
+[here](http://releases.k8s.io/release-1.0/docs/proposals/deployment.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Deployment
+
+## Abstract
+
+A proposal for implementing a new resource - Deployment - which will enable
+declarative config updates for Pods and ReplicationControllers.
+
+Users will be able to create a Deployment, which will spin up
+a ReplicationController to bring up the desired pods.
+Users can also target the Deployment at existing ReplicationControllers, in
+which case the new RC will replace the existing ones. The exact mechanics of
+replacement depends on the DeploymentStrategy chosen by the user.
+DeploymentStrategies are explained in detail in a later section.
+
+## Implementation
+
+### API Object
+
+The `Deployment` API object will have the following structure:
+
+```go
+type Deployment struct {
+  TypeMeta
+  ObjectMeta
+
+  // Specification of the desired behavior of the Deployment.
+  Spec DeploymentSpec
+
+  // Most recently observed status of the Deployment.
+  Status DeploymentStatus
+}
+
+type DeploymentSpec struct {
+  // Number of desired pods. This is a pointer to distinguish between explicit
+  // zero and not specified. Defaults to 1.
+  Replicas *int
+
+  // Label selector for pods. Existing ReplicationControllers which have all these
+  // label selectors will be scaled down.
+  Selector map[string]string
+
+  // Describes the pods that will be created.
+  Template *PodTemplateSpec
+
+  // The deployment strategy to use to replace existing pods with new ones.
+  Strategy DeploymentStrategy
+}
+
+type DeploymentStrategy struct {
+  // Type of deployment. Can be "Recreate" or "RollingUpdate".
+  Type DeploymentType
+
+  // Rolling update config params. Present only if DeploymentType =
+  // RollingUpdate.
+  RollingUpdate *RollingUpdateDeploymentSpec
+}
+
+type DeploymentType string
+
+const (
+  // Kill all existing pods before creating new ones.
+  DeploymentRecreate DeploymentType = "Recreate"
+
+  // Replace the old RCs by new one using rolling update i.e gradually scale down the old RCs and scale up the new one.
+  DeploymentRollingUpdate DeploymentType = "RollingUpdate"
+)
+
+// Spec to control the desired behavior of rolling update.
+type RollingUpdateDeploymentSpec struct {
+  // The maximum number of pods that can be unavailable during the update.
+  // Value can be an absolute number (ex: 5) or a percentage of total pods at the start of update (ex: 10%)
+  // By default, a fixed value of 1 is used.
+  maxUnavailable IntOrString
```

Comment:
Recreate does have one additional use case we consider separate - we have a
mid deployment hook so you can run a schema migration mid flight.  Also,
when graceful delete lands terminating pods are not counted towards RC
status - so rolling won't guarantee that you touch base on 0 pods running.

On Aug 11, 2015, at 10:22 PM, Brian Grant <notifications@github.com> wrote:

In docs/proposals/deployment.md
<https://github.com/GoogleCloudPlatform/kubernetes/pull/12236#discussion_r36821587>
:

> +type DeploymentType string
> +
> +const (
> +  // Kill all existing pods before creating new ones.
> +  DeploymentRecreate DeploymentType = "Recreate"
> +
> +  // Replace the old RCs by new one using rolling update i.e gradually scale down the old RCs and scale up the new one.
> +  DeploymentRollingUpdate DeploymentType = "RollingUpdate"
> +)
> +
> +// Spec to control the desired behavior of rolling update.
> +type RollingUpdateDeploymentSpec struct {
> +  // The maximum number of pods that can be unavailable during the update.
> +  // Value can be an absolute number (ex: 5) or a percentage of total pods at the start of update (ex: 10%)
> +  // By default, a fixed value of 1 is used.
> +  maxUnavailable IntOrString

FWIW, Recreate can be expressed as maxUnavailable = 100%.

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/pull/12236/files#r36821587>
.